File: fix_all.js
const fs = require('fs');
const p = 'swarm/src-tauri/src/lib.rs';
let lines = fs.readFileSync(p, 'utf8').split('\n');

const IND = Array(12).fill(' ').join('');

// Strategy: find lines by content, replace by index

// 1. Find "let shell = std::env::var" line and insert launchctl before it
let shellIdx = null;
for (let j = 0; j < lines.length; j++) {
 if (lines[j].includes('let shell = std::env::var("SHELL")')) {
 shellIdx = j;
 break;
 }
}

if (shellIdx === null) { console.log('ERROR: shell line not found'); process.exit(1); }

const block = [
 IND + '// On macOS, ask launchd for the GUI PATH first (no shell probe needed)',
 IND + 'if cfg!(target_os = "macos") {',
 IND + 'if let Ok(lc) = std::process::Command::new("launchctl")',
 IND + '.args(["getenv", "PATH"])',
 IND + '.output()',
 IND + '{',
 IND + 'if lc.status.success() {',
 IND + 'let raw = String::from_utf8_lossy(&lc.stdout).trim().to_string();',
 IND + 'if !raw.is_empty() {',
 IND + 'let dirs: Vec<PathBuf> = std::env::split_paths(&raw).collect();',
 IND + 'if !dirs.is_empty() { return dirs; }',
 IND + '}',
 IND + '}',
 IND + '}',
 IND + '}',
 IND + '',
];

lines.splice(shellIdx, 0, ...block);
console.log('Inserted launchctl block at line', shellIdx + 1);

// 2. Now find and fix .arg("-ilc") and .arg("echo -n $PATH")
let ilcIdx = null;
for (let j = 0; j < lines.length; j++) {
 if (lines[j].includes('"-ilc"')) {
 ilcIdx = j;
 break;
 }
}
if (ilcIdx !== null) {
 lines[ilcIdx] = IND + ' .arg("-lc")';
 console.log('Fixed -ilc at line', ilcIdx + 1);
}

let echoIdx = null;
for (let j = 0; j < lines.length; j++) {
 if (lines[j].includes('echo -n $PATH')) {
 echoIdx = j;
 break;
 }
}
if (echoIdx !== null) {
 lines[echoIdx] = IND + ' .arg("printf \'%s\' \\"$PATH\\"")';
 console.log('Fixed echo-n at line', echoIdx + 1);
}

// 3. Find and fix match block lines
let pathStrIdx = null;
for (let j = 0; j < lines.length; j++) {
 if (lines[j].includes('let path_str = String::from_utf8_lossy(&out.stdout)')) {
 pathStrIdx = j;
 break;
 }
}
if (pathStrIdx !== null) {
 lines[pathStrIdx] = IND + ' let raw_str = String::from_utf8_lossy(&out.stdout).trim().to_string();';
 lines[pathStrIdx + 1] = IND + ' if !raw_str.is_empty() { return std::env::split_paths(&raw_str).collect(); }';
 console.log('Fixed raw_str at line', pathStrIdx + 1);
}

let exitedIdx = null;
for (let j = 0; j < lines.length; j++) {
 if (lines[j].includes('login shell PATH probe exited non-zero:') && lines[j].includes('{:?}')) {
 exitedIdx = j;
 break;
 }
}
if (exitedIdx !== null) {
 lines[exitedIdx] = IND + ' println!("[Rust] login shell PATH probe exited non-zero");';
 console.log('Fixed exited non-zero at line', exitedIdx + 1);
}

let failedIdx = null;
for (let j = 0; j < lines.length; j++) {
 if (lines[j].includes('login shell PATH probe failed to run:') && lines[j].includes('{e:?}')) {
 failedIdx = j;
 break;
 }
}
if (failedIdx !== null) {
 lines[failedIdx] = IND + ' println!("[Rust] login shell PATH probe failed: {e:?}");';
 console.log('Fixed failed to run at line', failedIdx + 1);
}

// 4. Add extra bin dirs
let antigIdx = null;
for (let j = 0; j < lines.length; j++) {
 if (lines[j].includes('.antigravity-ide') && lines[j].includes('dirs.push')) {
 antigIdx = j;
 break;
 }
}

if (antigIdx !== null) {
 const extraDirs = [
 IND + 'dirs.push(PathBuf::from(&home).join(".claude").join("bin"));',
 IND + 'dirs.push(PathBuf::from(&home).join(".cursor").join("bin"));',
 IND + 'if let Ok(nvm) = std::env::var("NVM_DIR") {',
 IND + 'dirs.push(PathBuf::from(&nvm).join("versions").join("node").join("v").join("bin"));',
 IND + 'dirs.push(PathBuf::from(&nvm).join("versions").join("node").join(".bin"));',
 IND + '}',
 IND + 'if let Ok(volta) = std::env::var("VOLTA_HOME") {',
 IND + 'dirs.push(PathBuf::from(&volta).join("bin"));',
 IND + '}',
 IND + 'if let Ok(fnm) = std::env::var("FNM_DIR") {',
 IND + 'dirs.push(PathBuf::from(&fnm).join("bin"));',
 IND + '}',
 IND + 'dirs.push(PathBuf::from(&home).join(".asdf").join("shims"));',
 ];
 lines.splice(antigIdx + 1, 0, ...extraDirs);
 console.log('Inserted extra dirs at line', antigIdx + 2);
}

fs.writeFileSync(p, lines.join('\n'));
console.log('\nAll fixes applied!');

File: fix_all.rb
#!/usr/bin/env ruby
file = 'swarm/src-tauri/src/lib.rs'
content = File.read(file)

# Fix 1: -ilc -> -lc
content = content.gsub('.arg("-ilc")', '.arg("-lc")')

# Fix 2: echo -n $PATH -> printf '%s' "$PATH"
content = content.gsub('.arg("echo -n $PATH")', '.arg("printf \'%s\' \\"$PATH\\"")')

# Fix 3: path_str -> raw_str
content = content.gsub('let path_str = String::from_utf8_lossy(&out.stdout)', 'let raw_str = String::from_utf8_lossy(&out.stdout)')

# Fix 4: raw_str guard with proper indentation
content = content.gsub(
 /std::env::split_paths\(&path_str\)\.collect\(\)/,
 'if !raw_str.is_empty() { return std::env::split_paths(&raw_str).collect(); }\n Vec::new()'
)

# Fix 5: Error messages
content = content.gsub(
 'println!("[Rust] login shell PATH probe exited non-zero: {:?}", out.status);',
 'println!("[Rust] login shell PATH probe exited non-zero");'
)
content = content.gsub(
 'println!("[Rust] login shell PATH probe failed to run: {e:?}");',
 'println!("[Rust] login shell PATH probe failed: {e:?}");'
)

# Fix 6: Insert launchctl block before "let shell"
launchctl_block = <<~BLOCK
 // On macOS, ask launchd for the GUI PATH first (no shell probe needed)
 if cfg!(target_os = "macos") {
 if let Ok(lc) = std::process::Command::new("launchctl")
 .args(["getenv", "PATH"])
 .output()
 {
 if lc.status.success() {
 let raw = String::from_utf8_lossy(&lc.stdout).trim().to_string();
 if !raw.is_empty() {
 let dirs: Vec<PathBuf> = std::env::split_paths(&raw).collect();
 if !dirs.is_empty() { return dirs; }
 }
 }
 }
 }

BLOCK

content = content.sub(
 /(\s+)let shell = std::env::var\("SHELL"\)/,
 launchctl_block + '\1let shell = std::env::var("SHELL")'
)

File.write(file, content)
puts "All fixes applied!"
